### ISSUE TITLE
using index on multiple range conditions fails

### ISSUE THREAD

**Issue by algarsi3** [neo4j/neo4j#12042]
- Neo4j version: 3.4.7
- Operating system: Windows 10
- API/Driver: Cypher
- **Steps to reproduce**  
Create 3 nodes with 1 numeric attribute:
```
create (:person {number: 1})
create (:person {number: 2})
create (:person {number: 3})
```
Create an index on the attribute:
```
create index on :person(number)
```
Profile a query that uses NodeIndexSeekByRange with 1 condition
```
profile 
match (p:person)
where 0 <= p.number <= 1
return p
```
Again, profile a query that uses NodeIndexSeekByRange, but with 2 conditions
```
profile 
match (p:person)
where (0 <= p.number <= 1 or 3 <= p.number <= 4)
return p
```
Here the planner does not use NodeIndexSeekByRange but NodeLabelByScan. If we force it to use the index:
```
profile 
match (p:person)
using index p:person(number)
where (0 <= p.number <= 1 or 3 <= p.number <= 4)
return p
```
- Expected behavior
It uses the index.
- Actual behavior
Throws an error.
```
Neo.DatabaseError.Statement.ExecutionFailed
Neo.DatabaseError.Statement.ExecutionFailed: Failed to fulfil the hints of the query.
Could not solve these hints: `USING INDEX p:person(number)`

Plan Selection(ListBuffer(Ors(Set(Ands(Set(GreaterThanOrEqual(Property(Variable(p),PropertyKeyName(number)),Parameter(  AUTOINT1,Integer)), LessThanOrEqual(Property(Variable(p),PropertyKeyName(number)),Parameter(  AUTOINT0,Integer)))), Ands(Set(GreaterThanOrEqual(Property(Variable(p),PropertyKeyName(number)),Parameter(  AUTOINT3,Integer)), LessThanOrEqual(Property(Variable(p),PropertyKeyName(number)),Parameter(  AUTOINT2,Integer)))))))) {
  LHS -> NodeByLabelScan(p, LabelName(person), Set()) {}
}
```
I do not know if this is intended. I might be wrong but I expected the index to be used regardless the number the conditions over the index.

**Comment by SaschaPeukert** [comment]
Hej @algarsi3 ,

Thanks for reporting. We will look into this.

**Comment by fickludd** [comment]
@algarsi3 I agree with your sentiment that the planner should understand to use the index to perform two range seeks in this case. As the query does not return the wrong results however, we will not fix this in the 3.4 series, but defer fixing it until a later version (3.5 is close to being released, so it will not make that series either unfortunately).

You may close this issue if you want, or we can leave it open until it is fixed.

**Comment by boggle** [comment]
For this particular query, you can work around this issue using UNION.

**Comment by fickludd** [comment]
To elaborate, use `UNION` by doing:

```
MATCH (p:person)
WHERE 0 <= p.number <= 1
RETURN p
UNION
MATCH (p:person)
WHERE 3 <= p.number <= 4
RETURN p
```

**Comment by algarsi3** [comment]
Thanks for all the responses.

@fickludd, @boggle unfortunately my use case contains more complex logic and this is not a solution a the moment.

Has there been any progess regarding the issue?

**Comment by arnefischereit** [comment]
Hi @algarsi3,

Sorry for not coming back on this issue for such a long time.

We have logic in the planner now to automatically distribute OR-conditions across several UNION branches. I just tried your example

```cypher
MATCH (p:person)
  WHERE 0 <= p.number <= 1
     OR 3 <= p.number <= 4
RETURN p
```

and this is being planned the way that @fickludd described in https://github.com/neo4j/neo4j/issues/12042#issuecomment-431791060.

Does this now also solve your more complex case?

Regards,

Arne

**Comment by arnefischereit** [comment]
Hi @algarsi3,

We will now consider this issues being resolved. Should the problem persist or reappear, please reopen this issue or open a new one.

Regards,

Arne
